### PR TITLE
[PPC] Add DwarfRegAlias for VSRPair

### PR DESCRIPTION
**Pull request by syzaara**
Add DwarfRegAlias for VSRPair as it shares dwarfRegNum with the VR registers.

### CODE REVIEW

File: llvm/lib/CodeGen/StackMaps.cpp [hunk 1]
```diff
@@ -191,14 +191,22 @@ unsigned StackMaps::getNextMetaArgIdx(const MachineInstr *MI, unsigned CurIdx) {
...
+/// Go up the super-register and sub-register chain until we hit a valid dwarf
+/// register number.
 static unsigned getDwarfRegNum(unsigned Reg, const TargetRegisterInfo *TRI) {
   int RegNum;
   for (MCPhysReg SR : TRI->superregs_inclusive(Reg)) {
     RegNum = TRI->getDwarfRegNum(SR, false);
     if (RegNum >= 0)
       break;
   }
+  if (RegNum < 0) {
```
Comment (bzEq) [inline llvm/lib/CodeGen/StackMaps.cpp:203]: Is this change still necessary after using `DwarfRegAlias`? Looks we have similar register structure as https://github.com/llvm/llvm-project/commit/ddffa0e160036229f19e0ec9867952fc4c4dd66c.

File: llvm/test/CodeGen/PowerPC/ppc64-anyregcc.ll [hunk 2]
```diff
@@ -1,4 +1,4 @@
-; RUN: llc -verify-machineinstrs < %s | FileCheck %s
```
Comment (bzEq) [inline llvm/test/CodeGen/PowerPC/ppc64-anyregcc.ll:1]: Maybe add another `RUN` for `pwr10`, not update this one.

File: llvm/lib/CodeGen/StackMaps.cpp (hunk 1 above)
Comment (chenzheng1030) [reply llvm/lib/CodeGen/StackMaps.cpp:203]: I think for this issue, adding `DwarfRegAlias` should be enough to solve the missing DWARF number assertion for PPC pair vsr registers.

Maybe we can get a more general solution here. This is a static `getDwarfRegNum()` in `StackMap.cpp`. There is another implementation for `getDwarfRegNum()` in `MCRegisterInfo.h`.
- We should try to replace this static implementation with the one in `MCRegisterInfo.h`?
- Do we really need the `DwarfRegAlias` in td files? For the current implementation of `getDwarfRegNum()` in `MCRegisterInfo.h`, yes, we need `DwarfRegAlias`. But if we change its implementation to also check the sub/super registers, I don't think we still need them. The implementation seems make sense at least on PPC. For PPC registers, if registers overlap physically, they will share same DWARF register number.  (For now, most usages of `DwarfRegAlias` are for super/sub registers, several usages should be able to replace with DwarfRegNum, or register `Aliases`). For example:

PPC:
```
// The representation of r0 when treated as the constant 0.
let isConstant = true in {
def ZERO  : GPR<0, "0">,    DwarfRegAlias<R0>; 
def ZERO8 : GP8<ZERO, "0">, DwarfRegAlias<X0>;
} // isConstant = true
```
Can be replaced with:
```
// The representation of r0 when treated as the constant 0.
let isConstant = true in {
def ZERO  : GPR<0, "0"> {
  let Aliases = [R0]
} 
def ZERO8 : GP8<ZERO, "0"> {
  let Aliases = [X0]
}
} // isConstant = true
```

AArch64:
```
let isConstant = true in
def WZR   : AArch64Reg<31, "wzr">, DwarfRegAlias<WSP>;  //can be replaced with DwarfRegNum, can it be replaced with register Aliases?
```

One thing is, the register with DWARF number in td files may not be the direct super/sub register of the register we want to get the DWARF number for.

@arsenm @efriedma-quic any thoughts for this proposal?

Comment (arsenm) [reply llvm/lib/CodeGen/StackMaps.cpp:203]: Definitely should try to just use the MCRegisterInfo version instead of reconstructing it. 


> Do we really need the DwarfRegAlias in td files? 

I don't know, but the rest of your comment implies to me it only works if you can't address all parts of the register. e.g. if you have sub0_sub1, a direct sub1 use should not have the same number as sub0_sub1 (but maybe sub0 could have the same as sub0_sub1)? I haven't really looked at subregisters in dwarf before

Comment (chenzheng1030) [reply llvm/lib/CodeGen/StackMaps.cpp:203]: OK. Thanks for comment @arsenm 

I am fine if we solve PPC's issue by just adding `DwarfRegAlias` in PPC td files. We can do the refactor for the two versions `getDwarfRegNum()` in another patch.

**Review by chenzheng1030 — APPROVED** [review]
LGTM.

I created https://github.com/llvm/llvm-project/issues/96129 for the follow ups related to refactor of `getDwarfRegNum()`
